src/librustc_lint/builtin.rs

```diff
@@ -2153,123 +2153,208 @@ impl ClashingExternDeclarations {
-            // Do a full, depth-first comparison between the two.
-            use rustc_middle::ty::TyKind::*;
-            let a_kind = &a.kind;
-            let b_kind = &b.kind;
-
-            let compare_layouts = |a, b| -> bool {
-                let a_layout = &cx.layout_of(a).unwrap().layout.abi;
-                let b_layout = &cx.layout_of(b).unwrap().layout.abi;
-                debug!("{:?} == {:?} = {}", a_layout, b_layout, a_layout == b_layout);
-                a_layout == b_layout
-            };
+        // In order to avoid endlessly recursing on recursive types, we maintain a "seen" set.
+        // We'll need to store every combination of types we encounter anyway, so we also memoize
+        // the result.
+        struct SeenSet<'tcx>(FxHashMap<(Ty<'tcx>, Ty<'tcx>), Option<bool>>);
+
+        enum SeenSetResult {
+            /// We've never seen this combination of types.
+            Unseen,
+            /// We've seen this combination of types, but are still computing the result.
+            Computing,
+            /// We've seen this combination of types, and have already computed the result.
+            Computed(bool),
+        }
+
+        impl<'tcx> SeenSet<'tcx> {
+            fn new() -> Self {
+                SeenSet(FxHashMap::default())
+            }
+            /// Mark (a, b) as `Computing`.
+            fn mark_computing(&mut self, a: Ty<'tcx>, b: Ty<'tcx>) {
+                self.0.insert((a, b), None);
+            }
+            /// Mark (a, b) as `Computed(result)`.
+            fn mark_computed(&mut self, a: Ty<'tcx>, b: Ty<'tcx>, result: bool) {
+                *self.0.get_mut(&(a, b)).expect("Missing prior call to mark_computing") =
+                    Some(result);
+            }
+            fn get(&self, a: Ty<'tcx>, b: Ty<'tcx>) -> SeenSetResult {
+                match self.0.get(&(a, b)) {
+                    None => SeenSetResult::Unseen,
+                    Some(None) => SeenSetResult::Computing,
+                    Some(Some(b)) => SeenSetResult::Computed(*b),
+                }
+            }
+        }
+        fn structurally_same_type_impl<'tcx>(
+            seen_types: &mut SeenSet<'tcx>,
+            cx: &LateContext<'tcx>,
+            a: Ty<'tcx>,
+            b: Ty<'tcx>,
+            ckind: CItemKind,
+        ) -> bool {
+            debug!("structurally_same_type_impl(cx, a = {:?}, b = {:?})", a, b);
+            match seen_types.get(a, b) {
+                // If we've already computed the result, just return the memoized result.
+                SeenSetResult::Computed(result) => result,
+                // We are already in the process of computing structural sameness for this type,
+                // meaning we've found a cycle. The types are structurally same, then.
+                SeenSetResult::Computing => true,
+                // We haven't seen this combination of types at all -- compute their sameness.
+                SeenSetResult::Unseen => {
```

Comment:
```suggestion
            match seen_types.get(a, b) {
                // If we've already computed the result, just return the memoized result.
                SeenSetResult::Computed(result) => return result,
                // We are already in the process of computing structural sameness for this type,
                // meaning we've found a cycle. The types are structurally same, then.
                SeenSetResult::Computing => return true,
                // We haven't seen this combination of types at all -- compute their sameness.
                SeenSetResult::Unseen => (),
            }
```
nit: This function has a fairly large indentation, so using this match as an early exit would be cleaner imo